docs/weave-tutorial/03-codec.md

```diff
@@ -0,0 +1,219 @@
+}
+```
+
+This is a productive workflow and I recommend trying it out. You may find it doesn’t work for you and you can try other approaches, like copying the protobuf generated structs into some custom-written structs you like and then copying back into protobuf structs for serialization. You can also try playing with special [protobuf extensions](https://github.com/gogo/protobuf/blob/master/extensions.md) flags in your protobuf files to shape the auto-generated code into the exact shape you want.
+
+## Notes About oneof
+
+**oneof** is a powerful feature to produce union/sum types in your protobuf structures. For example, you may have a public key which may be one of many different algorithms and can define cases for each, which can be switched upon in runtime. We also use this for the transaction to enumerate a set of possible messages that can be embedded in the transaction. A transaction may have any one of them and serialize and deserialize properly. Type-safety is enforced in compile-time and we can switch on the kind on runtime. (Example from [bcp-demo](https://github.com/iov-one/bcp-demo/blob/master/app/codec.proto)):
+
+```protobuf
+oneof sum {
+  cash.SendMsg send_msg = 1;
+  namecoin.CreateTokenMsg new_token_msg = 2;
+  namecoin.SetWalletNameMsg set_name_msg = 3;
+  escrow.CreateMsg create_escrow_msg = 4;
+  escrow.ReleaseMsg release_escrow_msg = 5;
+  escrow.ReturnMsg return_escrow_msg = 6;
+  escrow.UpdatePartiesMsg update_escrow_msg = 7;
+}
+```
+
+The only problem is that the generated code is ugly to some people’s eyes. This lies in the fact that there is no clean way to express sum types in Go, and you have to force an interface with private methods to close the set of possible types. Although some people have been so revolted by this code that they preferred to [write their own serialization library](https://github.com/tendermint/go-amino "go-amino"), I would suggest just taking the breath and getting to know it. Here are the relevant pieces:
+
+```go
+type Tx struct {
+    // msg is a sum type over all allowed messages on this chain.
+    //
+    // Types that are valid to be assigned to Sum:
+    //  *Tx_SendMsg
+    //  *Tx_CreateTokenMsg
+    //  *Tx_SetNameMsg
+    //  *Tx_CreateMsg
+    //  *Tx_ReleaseMsg
+    //  *Tx_ReturnMsg
+    //  *Tx_UpdateEscrowMsg
+    Sum isTx_Sum `protobuf_oneof:"sum"`
+...
+}
+
+type isTx_Sum interface {
+    isTx_Sum()
+    MarshalTo([]byte) (int, error)
+    Size() int
+}
+
+type Tx_SendMsg struct {
+    SendMsg *cash.SendMsg `protobuf:"bytes,1,opt,name=send_msg,json=sendMsg,oneof"`
+}
+type Tx_CreateTokenMsg struct {
+    CreateTokenMsg *namecoin.CreateTokenMsg `protobuf:"bytes,2,opt,name=new_token_msg,json=newTokenMsg,oneof"`
+}
+```
+
+We now have some intermediate structs that give us a layer of indirection in order to enforce the fact we can now securely switch over all possible `tx.Sum` fields, with [code like this](https://github.com/iov-one/bcp-demo/blob/master/app/tx.go#L33-61):
```

Comment:
`I don't understand the word indirection`

Reply:
I feel `redirection` would fit best in that case